portal: Take the code of the portal to a different repo

psaiz opened the pull request:
Work In Progress.
Addresses https://github.com/cernopendata/opendata.cern.ch/issues/3589

This should not be merged untitl the cernopendata-portal repo is public and the changes have been validated by our contributors.

tiborsimko commented on .github/workflows/ci.yml:
```diff
@@ -118,49 +41,3 @@ jobs:
 
       - name: Check fixtures
         run: ./run-tests.sh --check-fixtures
```
Note that we also spoke about strict schema checking both on the DB side and the OpenSearch side. Have you created the checker? Can you add it to the `run-tests.sh` and the CI?  We should check any troubles early.

psaiz replied:
It would be better to add that new functionality later on. This PR is already big enough.

psaiz commented on .github/workflows/ci.yml:
```diff
@@ -25,83 +25,6 @@ name: CI
 on: [push, pull_request]
 
 jobs:
-  lint-shellcheck:
```
ok, done

psaiz commented on .github/workflows/ci.yml:
```diff
@@ -25,83 +25,6 @@ name: CI
-          sudo apt-get install shellcheck
-          ./run-tests.sh --check-shellscript
-
-  lint-black:
```
ok, done

psaiz commented on .github/workflows/ci.yml:
```diff
@@ -25,83 +25,6 @@ name: CI
-          pip install black
-          ./run-tests.sh --check-black
-
-  lint-pycodestyle:
```
Ok, done. Note that pycodestyle was only checking the cernopendata directory, so making it check the scripts directory will be a change of behaviour

psaiz commented on .github/workflows/ci.yml:
```diff
@@ -25,83 +25,6 @@ name: CI
-          pip install pycodestyle
-          ./run-tests.sh --check-pycodestyle
-
-  lint-pydocstyle:
```
Ok, done. Note that pydocstyle was only checking the cernopendata directory, so making it check the scripts is a change a of behaviour

psaiz commented on .github/workflows/ci.yml:
```diff
@@ -25,83 +25,6 @@ name: CI
-          pip install pydocstyle
-          ./run-tests.sh --check-pydocstyle
-
-  lint-check-manifest:
```
Remoiving the MANIFEST

psaiz commented on .github/workflows/ci.yml:
```diff
@@ -118,49 +41,3 @@ jobs:
       - name: Check fixtures
         run: ./run-tests.sh --check-fixtures
-
-  check-isort:
```
ok, done

psaiz commented on DEVELOPING.rst:
```diff
@@ -8,461 +8,91 @@
-context (using ``docker-compose.yml``), you can use a helper script joining the
-above tips together to quickly initialise your working environment.
+At this point, all the services should be up and running. If you go to a web browser to http://0.0.0.0:5000/, you should
+see the web portal, with a few entries inside (the vocabularies, and news about the portal itself).
```
What terminology do you suggest for something that includes records/documents/glossary terms? Specifying the three would be too much.

tiborsimko replied:
I would keep the three sections separated, as is the case now. When someone work with docs, we need to explain how to use Markdown, use LaTeX, how to add static images, etc. These could live under the documentation sertion, because we don't support Markdown in records, etc. Just an example why these were separated sections before. (BTW you have removed those parts?)

As for the terms, these are living in the `cernopendata-portal` repository, so we should just point there how to add terms, load, check, etc. So the guide won't be detailed here.  (Although IMO the glossary terms should live with the content, not with the infrastructure.)

psaiz commented on DEVELOPING.rst:
```diff
@@ -8,461 +8,91 @@
+  $ # and the web container starts by setting up the development environment
+  $ sleep 120
+  $ # Create the basic structure
+  $ my_docker exec -it web /code/scripts/populate-instance.sh
```
This is populating the part that comes from the infrastructure: glossary terms and a handful of documents.  I would keep that, it doesn't take that much time

tiborsimko replied:
Note that the script fails:

```
$ time my_docker exec -it web /code/scripts/populate-instance.sh
...
docs simulated-dataset-categories inserted
/usr/lib64/python3.9/site-packages/XRootD/client/finalize.py:46: DeprecationWarning: Importing 'itsdangerous.json' is deprecated and will be removed in ItsDangerous 2.1. Use Python's 'json' module instead.
  if isinstance(obj, File) and obj.is_open():

ls: cannot access 'cernopendata/modules/fixtures/data/records/*.json': No such file or directory
docker compose -f ../cernopendata-portal/docker-compose.yml -f  exec -it web   0.08s user 0.06s system 0% cpu 56.730 total
```

psaiz replied:
Thanks for pointing that out. Since that comes from the `cernopendata_portal`, it should be fixed there (and with a different PR).

psaiz commented on DEVELOPING.rst:
```diff
@@ -8,461 +8,91 @@
-
-The following sections document the development life cycle in fuller
-detail.
+   $ my_docker exec -it web /content/scripts/load-fixtures.sh
```
The way I see it, this one will be hardly ever used. We should encourage people to load only the entries that they are modifying (as explained in point 4). For the cases where we do want to recreate the full instance from scratch, having this script was useful.

tiborsimko replied:
Fully agreed, this is why the current guide is using things like:

```
$ docker exec -i -t opendatacernch-web-1 cernopendata fixtures records -f /code/cernopendata/modules/fixtures/data/records/opera-author-list-multiplicity.json --mode insert
```

WRT point 4, it's always useful to use a copy-paste-ready example, e.g. `full_path` can point to a Markdown file, and people may wonder why it's not working.

WRT loading all fixtures, shouldn't the `cernopendata fixtures` do that already? It seems to "promise" loading all fixtures by default:

```console
$  my_docker exec -it web cernopendata fixtures records --help

Usage: cernopendata fixtures records [OPTIONS]

  Load all records.

Options:
  --skip-files                    Skip loading of files
  -f, --file PATH                 Path to the file(s) to be loaded. If not
                                  provided, allfiles will be loaded
  --profile                       Output profiling information.
  --mode [insert|replace|insert-or-replace]
                                  [required]
  --help                          Show this message and exit.
```

psaiz replied:
ok, removed

psaiz commented on DEVELOPING.rst:
```diff
@@ -8,461 +8,91 @@
-       fetch = +refs/heads/*:refs/remotes/upstream/*
-       fetch = +refs/pull/*/head:refs/remotes/upstream/pr/*
-
-Understanding repository branches
```
ok, done

psaiz commented on DEVELOPING.rst:
```diff
@@ -8,461 +8,91 @@
-Understanding repository branches
----------------------------------
-
-We use three official base branches:
```
ok, done. I've changed a couple of things in the description:
1. s/feature/content
2. Add the merge into the development, which was not mentioned

tiborsimko replied:
I would prefer to keep the `git` examples as well, to show the real snippets of typical work, but I can update the documentation also after the merge. (Including bits like `--skip-docs` and friends for fast development that were removed in this PR.)

psaiz commented on DEVELOPING.rst:
```diff
@@ -8,461 +8,91 @@
 
-   $ git push origin master
-   $ git push origin :improve-event-display-icons
+   $ my_docker exec -it web /data/scripts/load-images.sh
```
done, thanks

tiborsimko replied:
I tested that the images are well copied, for example:

```
[invenio@d2d5c9b658be code]$ ls -l /content/data/images/subdir/logo-iris-hep.png
-rw-r--r-- 1 invenio 1000 12901 Aug 13 07:21 /content/data/images/subdir/logo-iris-hep.png
```

However the existing content images are still located in the traditional place:

```
[invenio@d2d5c9b658be code]$ ls -l /code/cernopendata/modules/theme/static/docs/cms-releases-heavy-ion-data/cms_hi_event_1.png
-rw-r--r-- 1 invenio 1000 288993 Feb 27 17:36 /code/cernopendata/modules/theme/static/docs/cms-releases-heavy-ion-data/cms_hi_event_1.png
```

Two questions:

1. How are the copied user-supplied images exposed over the web? Under which URL?
2. It would be good to keep all the existing content images such as `cms_hi_event_1.png` here in the new directory as well, as suggested above. In this way, the users would have complete example how to add a Markdown docs with embedded images in one place.

psaiz commented on DEVELOPING.rst:
```diff
@@ -8,461 +8,91 @@
 .. code-block:: console
 
-   http://github.com/johndoe/opendata.cern.ch
+   $ my_docker exec -it web /content/scripts/check_fixtures.py
```
ok, done

psaiz commented on DEVELOPING.rst:
```diff
@@ -8,461 +8,91 @@
+   $ my_docker exec -it web /data/scripts/load-images.sh
 
-This would conclude your work on ``improve-event-display-icons``.
+..
```
It would be more natural if the description of "how to add a new field" is in the DEVELOPING guide of the repository that offers the schema, isn't it? From this one, we can add a link to that developing guide

tiborsimko replied:
Sure, such a link would perfectly do. My main aim was to not leave this subject topic "empty" here.

psaiz replied:
Isn't the link in line 59 enough for that? That mentions as well the schema.

tiborsimko replied:
That link seems to point just to the `cernopendata-portal` repository home, which does not mention much about how to add new field or a new mapping.  (Recall the old user story "how do I add a new magnet polarity field".) Sure, we can write these stories later, but it should not be forgotten... I thought the split is a perfect occasion to do that.

(BTW https://github.com/cernopendata/cernopendata-portal/blob/main/DEVELOPING.rst needs updating too.)

psaiz replied:
The updates to the cernopendata-portal will come with a dedicated PR there

psaiz commented:
Ok, I've put the images, and left there also the ones from DELPHI, so that things like https://github.com/cernopendata/cernopendata-portal/pull/47 are not needed anymore. Let me know if that addresses the concerns about the images.

tiborsimko commented on DEVELOPING.rst:
```diff
@@ -8,461 +8,91 @@
+  $ git clone https://github.com/cernopendata/cernopendata-portal.git
+  $ # Move to the directory of the content
+  $ cd opendata.cern.ch
+  $ # The command line is a bit long: let's create an alias to avoid typing all of that
```
Coming back to this resolved comment.  (BTW it may be good to keep them opened for easy search/find.) 

We could take advantage of [includes and overrides](https://docs.docker.com/compose/multiple-compose-files/include/#include-and-overrides) to create the following file `docker-compose.yml` file in this repository:

```yaml
include:
  - path:
      - ../cernopendata-portal/docker-compose.yml
      - docker-compose-content.yml
```

This should allow people to type directly the usual `docker compose` commands they are already used to, such as:

```console
$ docker compose pull
$ docker compose up
$ docker exec -i -t opendatacernch-web-1 cernopendata fixtures records \
              --mode insert-or-replace \
              -f /content/data/records/cms-primary-datasets.json
```

This should be more comfortable user-experience-wise when compared to introducing an unusual `my_docker` alias that may look a bit strange.
 
(I would also rename `docker-compose-content.yml` to something like `docker-compose-override.yml` to make its purpose very clear.)

tiborsimko replied:
@psaiz Unresolving this conversation because `my_docker` still seems to be there?

psaiz replied:
Please, submit a suggestion with the exact changes that you want. That will speed things up.

tiborsimko replied:
It is not possible to send suggestion in the GitHub style, because I'm proposing to alter files. If my above comment was not clear, then this is what I was proposing:

```bash
echo "include:" > docker-compose.yml
echo " - path:" >> docker-compose.yml
echo "   - ../cernopendata-portal/docker-compose.yml" >> docker-compose.yml
echo "   - docker-compose-override.yml" >> docker-compose.yml
git add docker-compose.yml
git mv docker-compose-content.yml docker-compose-override.yml
git commit --amend --no-edit
```

The help instructions can simply say:

```console
$ docker compose pull
$ docker compose up
$  docker exec -i -t opendatacernch-web-1 /code/scripts/populate-instance.sh --skip-records --skip-glossary --skip-docs
$ docker exec -i -t opendatacernch-web-1 cernopendata fixtures records \
              --mode insert-or-replace \
              -f /content/data/records/cms-primary-datasets.json
```

This will do for now for the merge to happen. 

I'll be amending the documenation page considerably after merge, as I mentioned elsewhere (e.g. [here](https://github.com/cernopendata/opendata.cern.ch/pull/3590#discussion_r1714983018) and [here](https://github.com/cernopendata/opendata.cern.ch/pull/3590#discussion_r1746732736)) which were closed by not addressed. It will be simpler if I take care of that.


P.S.  Note that you'll need to also rebase against latest `master`.

psaiz replied:
Done, and rebased
